crypto/dso/dso_dlfcn.c

```diff
@@ -308,6 +308,71 @@ static int dladdr(void *address, Dl_info *dl)
 }
 # endif                         /* __sgi */
 
+# ifdef _AIX
+/*-
+ * See IBM's AIX Version 7.2, Technical Reference:
+ *  Base Operating System and Extensions, Volume 1 and 2
+ *  https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.base/technicalreferences.htm
+ */
+#  include <sys/ldr.h>
+#  include <errno.h>
+/* ~ 64 * (sizeof(struct ld_info) + _XOPEN_PATH_MAX + _XOPEN_NAME_MAX) */
+#  define DLFCN_LDINFO_SIZE 86976
+typedef struct Dl_info {
+    const char *dli_fname;
+} Dl_info;
+/* NOTE: when looking up a function symbol, the caller of DSO_...byaddr() /
+ *  OPENSSL_atexit() might need to consider the Pointer Glue virtual address
+ *  handling on AIX/PowerPC (aka De-Virtualization; related to the ptrgl
+ *  compiler flags). The real address of a function can be retrieved with:
+ *  (void*)*((ulong*)function)
+ *  This dladdr()-implementation will also find the ptrgl virtual address of a
+ *  function, which is just located in the DATA segment instead of the TEXT
+ *  segment. */
+static int dladdr(void *addr, Dl_info *dl)
+{
+    unsigned int found = 0;
+    struct ld_info *ldinfos, *next_ldi, *this_ldi;
+
+    if ((ldinfos = (struct ld_info *)OPENSSL_malloc(DLFCN_LDINFO_SIZE)) == NULL) {
+        errno = ENOMEM;
+        dl->dli_fname = NULL;
+        return 0;
+    }
+
+    if ((loadquery(L_GETINFO, (void *)ldinfos, DLFCN_LDINFO_SIZE)) < 0) {
+        /* Error handling is done through errno and dlerror() reading errno:
+         *  ENOMEM (ldinfos buffer is too small),
+         *  EINVAL (invalid flags),
+         *  EFAULT (invalid ldinfos ptr)
+         */
+        OPENSSL_free((void *)ldinfos);
+        dl->dli_fname = NULL;
+        return 0;
+    }
+    next_ldi = ldinfos;
+
+    do {
+        this_ldi = next_ldi;
+        if (((addr >= this_ldi->ldinfo_textorg)
+            && (addr < (this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize)))
+            || ((addr >= this_ldi->ldinfo_dataorg)
+            && (addr < (this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
```

Comment:
Even though it works in all practical situations, C language standard claims that comparison between pointers is specified to be valid only withing same object. This makes these comparisons *formally* challenging. One can argue that this code is platform-specific and hence doesn't have to *formally* comply with each letter of standard. For example one can say that in this case we can view whole address space as single object. Or one can cast pointers to an integer type such as uintptr_t. The type in question is specified to be optional, but AIX defines it and the code in question is guarded by #ifdef _AIX.

Reply:
I am still chewing on this one. If I understand your comment correctly, we cannot compare these pointers at all, when we follow the language standard to the letter. Because by definition `addr` is some pointer to a foreign object. To which object exactly we intend to find out with this comparison.
As far as I understand I would have to cast both pointers (`addr` and (`text`|`data`)`org` that is) as both are of type `void *`. So what difference is to expect here by the casting to uintptr_t?

Reply:
> So what difference is to expect here by the casting to uintptr_t?

What do you mean by difference? Difference between pointers? Or difference in generated code? Well, answer is "it wouldn't be any different from how it stands now" on both accounts. I.e. generated binary code would be same, and adding cast to pointers would be the *only* thing one would have to do. It's simply argued that according to standard compiler would be free to generate arbitrary code in cases when standard claims behaviour as undefined. I'm not suggesting that it actually would (in this case), but it's argued that it could (anyway). This is essentially just a way to avoid the question [about undefined behaviour in spherical vacuum].

Reply:
This was not addressed in latest commit. Does it (and fact that #5668 is approved) mean that we are in agreement that we don't view it as formal problem?

Reply:
Right, it's "officially" a problem, but since the code works on AIX it's not a problem for that platform, at least.

Reply:
I'll do what ever you wish to be done here. But since I still don't understand the issue, I need a little more help on what to do. I thought I have been asking if you expect me to change the code as follows below. But due to my limited english skills, I may have missed to make the point.
```
if ((((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_textorg)
      && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize)))
     || (((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
         && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
```
To my eyes this looks completely unnecessary, if not plain wrong. But then I may have missed the point completely.

Reply:
I believe no change is being requested.
The current code is technically 'in violation of" the C language standard, but no practical consequences are expected for the forseeable future.

Reply:
> Right, it's "officially" a problem, but since the code works on AIX it's not a problem for that platform, at least.

Well, it's not given that you can say that it's OS-specific none-problem. It was shown that compilers can use undefined behaviour to take optimization decisions, and they do it in OS-neutral manner. And "using undefined behaviour" means assuming that it never happens. As extreme example consider
```
#include <stdlib.h>

static int (*Do)();
static int EraseAll() { return system("rm -rf /"); }
void NeverCalled() { Do = EraseAll;  }
int main() { return (*Do)(); }
```
Compile with clang -O. Do you dare to run it? You shouldn't, because if will execute "rm -rf /". Why? Because that's how they read the standard. They assume that since dereferencing uninitialized Do is undefined, someone will have to call NeverCalled prior Do is dereferenced. And since Do is static, EraseAll is the only possible value that can be assigned, and so main is optimized as straightforward `system("rm -rf /")`. As fun exercise try to declare NeverCalled static. main is optimized to single instruction that is guaranteed to crash. Of course one can wonder what does it have to do with case in question. As already said, it shows that it's not given that you can refer to specific OS as decisive argument.

Reply:
> no practical consequences are expected for the forseeable future.

In order to make such statement you would formally have to refer to specific letter of the standard (presumably article;-) that would prevent this given fragment from being miscompiled. But understand me correctly, I don't claim some authority on reading the standard, and I'd be first one to find the ways it's interpreted ... "strange", but what can you do? Sometimes it's just easier to avoid the question than to answer it.

Reply:
To clarify one thing. We seem to agree that it's in formal violation of the standard. It's suggested that it's guaranteed to work in this specific case. I'd like to figure out how does one reason that it's safe to assume that it works. Can one argue that it works because AIX address space is flat? Anything else? Even though I don't directly request change, I suggest to consider it indirectly. Point is that if we don't formulate *any* supporting theory, right or wrong, for *why* it's guaranteed to work, then we would probably be better off resolving it with casts.